[v1.11] IPSec key rotation without agent restart

jibi opened the pull request:
* #19814 -- IPSec key rotation without agent restart (@jibi )
 * #19932 -- ipsec: fix stale keys reclaim logic (@jibi)
 * #20392 -- nodediscovery: make LocalNode return a deep copy of localNode (@jibi)

Moreover, as a dependency of the 2 PRs some additional commits have been backported from https://github.com/cilium/cilium/pull/18887:
  * https://github.com/cilium/cilium/commit/ed934cb958c52515be37998038301229959a0fd2
  * https://github.com/cilium/cilium/commit/9c1cb4856322a80174799dae142ef761e51e2310
  * https://github.com/cilium/cilium/commit/e52fe1d59d1c6f51a76993641ea9412e0f44f749

Once this PR is merged, you can update the PR labels via:
```upstream-prs
$ for pr in 19814 19932 20392 ; do contrib/backporting/set-labels.py $pr done 1.11; done
```
or with
```
$ make add-label BRANCH=v1.11 ISSUES=19814,19932,20392
```

gandro committed:
health: Move endpoint IP to node package

[ upstream commit ed934cb958c52515be37998038301229959a0fd2 ]

The Cilium health endpoint IP is the only static local node field which
is not accessible from the node package. To avoid code modifying the
`NodeDiscovery.LocalNode` field (which will become more dynamic in a
subsequent mode), the health IP is now moved to the node package with a
corresponding getter and setter like all other static node fields.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Gilberto Bertin <jibi@cilium.io>

gandro committed:
nodediscovery: Unify source of truth for node name

[ upstream commit 9c1cb4856322a80174799dae142ef761e51e2310 ]

Clustermesh related code is relying on
`NodeDiscovery.LocalNode.FullName()` to obtain the globally unique name
of the local node. This field is initialized from `nodeTypes.GetName()`.
This code removes all uses of `LocalNode.FullName()` and replaces it
with an equivalent call to `nodeTypes.GetAbsoluteNodeName()`. This
allows us to make the `NodeDiscovery.LocalNode` private in a subsequent
commit.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Gilberto Bertin <jibi@cilium.io>

gandro committed:
nodediscovery: Make LocalNode object private

[ upstream commit e52fe1d59d1c6f51a76993641ea9412e0f44f749 ]

The LocalNode currently consists of static fields which are never
changed after the local node has been initialized. However, with the
introduction of secondary allocation CIDRs, we want to be update the
secondary allocation CIDR fields dynamically at run-time.

To ensure writes to the object are always protected by a mutex, and that
to be able to ensure that any changes to the object are propagated to
relevant subsystems, this commit makes the `LocalNode` object private.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
nodediscovery: split StartDiscovery into smaller methods

[ upstream commit c19ef1c36ba890945507a59e3af642db8c743f27 ]

in preparation to a subsequent change, this commit moves some of the
(*NodeDiscovery).StartDiscovery() logic into separated methods:

* fillLocalNode(), responsible for syncing the localNode object with the
  actual state of the node
* updateLocalNode(), responsible for publishing the updated KV store
  entry and/or CiliumNode object for the local node

In addition to that, this commit introduces also the UpdateLocalNode()
method, which ensures that the proper lock is held while syncing the
localNode object state with fillLocalNode() and then publishing the
new object with updateLocalNode().

No functional changes are introduced.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
Move fswatcher package out of crypto/certloader

[ upstream commit d3b2d54d6f8c298478166c546155c0ab8ade497a ]

Given that this logic is totally agnostic to the crypto/certloader
package and in a subsequent commit it will be used also in the IPSec
subsystem, make it a top level package.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
nodediscovery: use time.After() instead of time.NewTimer().C

[ upstream commit b36cd0186e4c704514de17c23c763afab3caa8c1 ]

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
ipsec: implement keyfile watcher

[ upstream commit dce0d53359abd388bc6f72dd9134354482bf698a ]

This commit adds a new filesystem watcher for the IPSec subsystem, which
in turn will enable automatic key rotation without the need to restart
the agent whenever a change to the IPSec keyfile is detected.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

Conflicts:
* minor conflit in daemon/cmd/daemon.go as v1.10 does not have support
  for runtime device detection

jibi committed:
ipsec: fix stale keys reclaim logic

[ upstream commit 3b2e982bc8a131744aa403bc6616568794d8bfa5 ]

The current logic used to reclaim stale IPSec keys is incorrect. In
fact, whenever a new key is added we just spawn a new goroutine which
after linux_defaults.IPsecKeyDeleteDelay time will delete all keys
different than the one just added.

The issue with this approach is that if a new key is added while one of
these goroutine is still running, this goroutine will end up deleting
the new key.

This commit fixes this by introducing a single new goroutine that
periodically removes any stale key (XFRM state and XFRM policy) after
linux_defaults.IPsecKeyDeleteDelay has passed since the key was
replaced.

Fixes: #19814
Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
ipsec: rename ipSecKeysGlobalLock to ipSecLock

[ upstream commit 20cd921d8d35970c5cc58a71743122a899353c69 ]

as now this lock is used to protect also ipSecCurrentKeySPI in addition
to ipSecKeysGlobal.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
ipsec: add ipSecXfrmMark{Set,Get}SPI helpers

[ upstream commit b2331289c703156b5a965a4e488424353f1c0580 ]

which allows to encode and extract an SPI value in and from a XfrmMark

Conflicts:
* minor conflict in pkg/datapath/linux/ipsec/ipsec_linux.go as a bunch
  of constants are not defined in v1.10

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
logfields: add SPI and OldSPI IPSec constants

[ upstream commit 604ec254cca5057bac189a7b30e123399f6015a5 ]

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
nodediscovery: add LocalNode method

[ upstream commit 3d6ed0432ad94e0615c3417cb17fc487270331b8 ]

This method will be used in a subsequent commit to retrieve a copy of
the localNode object stored in the nodeDiscovery one.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
ipsec: replace 0 with FAMILY_ALL netlink constant

[ upstream commit d0ac0fa70de6931e83ffeb19a5838a964254c0b7 ]

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
docs: update IPSec docs

[ upstream commit 08d571a97ff6402c8654f99c4af42afbd25848a6 ]

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

julianwiedmann reviewed (approved):
ipsec looks good. As expected, code is essentially identical to latest upstream (module the parser diff in `loadIPSecKeys()`).

No surprises in the nodediscovery parts, but hopefully @gandro can have a closer look.

gandro commented on pkg/nodediscovery/nodediscovery.go:
```diff
@@ -302,6 +303,16 @@ func (n *NodeDiscovery) UpdateLocalNode() {
+	defer n.localNodeLock.Unlock()
+
+	n.fillLocalNode()
+	return n.localNode
```
I didn't notice this in the upstream PR, but don't we want to `DeepCopy()` here? `types.Node` contains a few slices that will be aliased. They are not modified after `StartDiscovery()` has been called in v1.10/v1.11, but they are in v1.12.

jibi replied:
Makes sense, in that case yes, we should switch to `DeepCopy()` :+1: 

I'll fill the master PR and backport the fix here and in the v1.10 one

gandro reviewed (approved):
Awesome work. Thank you especially for noting merge conflicts in the commits themselves, that was super helpful.

I focused on the nodediscovery part and that looks good (besides one minor thing also present on master). I only did a cursory glance at the IPSec changes.

jibi committed:
nodediscovery: make LocalNode return a deep copy of localNode

[ upstream commit 88e683528e92e20fe9557520a869030f234738a5 ]

as that object contains a few slices that will get modified after
StartDiscovery() is called

Fixes: 3d6ed0432a ("nodediscovery: add LocalNode method")
Suggested-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Gilberto Bertin <jibi@cilium.io>

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -40,6 +40,10 @@ const (
+
+	// ipSecXfrmMarkSPIShift defines how many bits the SPI is shifted when
+	// encoded in a XfrmMark
+	ipSecXfrmMarkSPIShift = 12
```
Commit `ipsec: add ipSecXfrmMark{Set,Get}SPI helpers` is missing a reference to the upstream commit.

pchaigno reviewed (approved):
One minor comment below, can probably be fixed without restarting the end-to-end tests afterward.
